poly/promise/promise.go

```diff
@@ -0,0 +1,744 @@
+		return n, err
+	}
+	
+	// Find the length of the signature index as stated above.
+	_, piLen := binary.Varint(buf)
+	if piLen <= 0 {
+		return 0, errors.New("Binary number not provided for index")
+	}
+	
+	// Retrieve the signature length
+	sigLen, sigLenBytesRead := binary.Varint(buf[piLen:])
+	if sigLenBytesRead <= 0 {
+		return 0, errors.New("Signature length not provided")
+	}
+
+	// The entire length of the buffer can now be calculated and the
+	// object = unmarshalled.
+	finalBuf := make([]byte, int(sigLen) + piLen + sigLenBytesRead)
+	
+	// Copy what has already been read into the buffer.
+	copy(finalBuf, buf)
+	
+	// Read the rest.
+	m, err2 := io.ReadFull(r, finalBuf[n:])
+	if err2 != nil {
+		return m, err2
+	}
+	return n+m, p.UnmarshalBinary(finalBuf)
+}
+
+/* PromiseShare is used to represent a secret share of a Promise. When a
+ * guardian wants to reveal the secret share it is guarding either to a promised
+ * client or to other guardians to prove that the promiser is crooked, it
+ * constructs this object.
+ *
+ * The key features of a PromiseShare are:
+ *	1. The index of the share
+ *      2. The share itself
+ *      3. The Diffie-Hellman secret between the guardian and promiser
+ *
+ * #3 is used for verification purposes. Other servers can use it to prove that
+ * encrypting #2 with #3 will produce the secret stored in the promise.
+ *
+ * As mentioned above, PromiseShare's can also be used as "BlameProof" objects.
+ *
+ * The BlameProof object provides an accountability measure. If a promiser
+ * decides to construct a faulty share, guardians can construct a BlameProof
+ * to show that the server is malicious. 
+ * 
+ * The guardian provides the index of the bad secret, the bad secret itself,
+ * and its diffie-hellman shared key with the promiser. Other servers can then
+ * verify if the promiser is malicious or the guardian is falsely accusing the
+ * server.
+ *
+ * To quickly summarize the blame procedure, two things must hold for the blame
+ * to succeed:
+ *
+ *   1. The provided share when encrypted with the diffie key must equal the
+ *   point provided at index bi of the promise. This verifies that the share
+ *   was actually intended for the guardian.
+ *
+ *   2. The provided share must fail to pass pubPoly.Check. This ensures that
+ *   the share is actually corrupted and the guardian is not just lying.
+ */
+
+type PromiseShare struct {
+
+	// The index of the share
+	i int
+	
+	// The actual share from the guardian
+	share abstract.Secret
+	
+	// The Diffie-Hellman key between guardian i and the promiser.
+	diffieKey abstract.Point
+}
+
+
+/* Initializes a new PromiseShare
+ *
+ * Arguments
+ *    i   = the index of the Promise share the guardian is revealing
+ *    s   = the share being revealed
+ *    d   = the Diffie-Hellman key between the guardian and promiser
+ *
+ * Returns
+ *   An initialized PromiseShare
+ */
+func (p *PromiseShare) Init(i int, s abstract.Secret, d abstract.Point) *PromiseShare {
```

Comment:
I think the issue is that I am combining network logic with crypto logic.

PromiseShare is more of a networking struct, since insurers will have to send the secret and the index over the network to clients who ask. I should move this code to prifi.

I've reworked promise.go to contain these structures:

    - PromiseState => Keeps track of signatures certifying the promise, secrets revealed,    
    and blame proofs received
 
    - PromiseSignature => a struct responsible for marshalling/ unmarshalling signature
    messages (along with keeping track of their suites). This makes it easier for
    PromiseState to keep track of signatures.

    - BlameProof => an object containing enough information to prove that a promiser
    created an invalid secret share